### DISCUSSION TITLE
Multiple Currencies

### DISCUSSION THREAD

**Discussion by SinsOfSolitude** [Zrips/CMI#3896] [Feature Requests]
**Description of feature/improvement request:**

Implementing the option to create multiple currencies and add per-world as well as mysql or CMIB support for them.

---

### Detailed explanation of what it should do, when and in what situation: ###

 You'd have the chance to create multiple currencies, define them as lets say "Silver Coins", "Gold Coins", "Gems", etc. You'd be able to set a custom currency exchange rate, like 1 Gem = 10 Gold Coins = 100 Silver Coins, all customizable.

 Per-world and bungeecord support via mysql OR CMIB would also be essential for many users. You would be able to set which currencies are enabled/disabled on which maps, what currencies are shared between servers over mysql, etc.

 This should also be integrated with the built in Worth system thats used for shops and command costs. You could set a placeholder for each currency, as currently command cost would be $1, with multiple currencies it could be defined as 1$1 then the next currency being 2$1, first number standing for which currency it is, second being the amount, possibly even with a decimal point. The currency type might not even need to be determined by a number, but could be just letters, or a different format overall, $1.00(GC), (GC) being the gold coins. There are a lot of options for this.

 Example of how this could be used. Theres a survival server and a hub server. Survival server has a survival world and shop world where only silver and gold coins are linked with both worlds, but gems exclusively on shop world. In survival you're able to earn silver coins for playtime/whatever else, gold coins for kills, etc. When you transfer to the shop, you can exchange the silver and gold coins to gems. Gems would be linked over bungeecord via mysql, each server's shop world will have all the gems updated. So when you go into the hub, you can use the gems to buy your cosmetic items, etc.

### Basic explanation as to why this feature is needed and/or useful to have: ###

 There haven't ever been a lot of plugins out there that could do multiple currencies, and the ones that there are, are all out-dated or abandoned from as far as I know. This would be insanely useful for scenarios like the example I wrote, Hub servers could have a special currency which can be gained by playing on sub-servers, different worlds could be different nations, theres just infinite options to what you could do with this, how this could be applied, and so on.

 With this feature, this would make CMI's economy system more advanced than any other standalone economy plugin out there, and I bet a ton of people would convert to this.

**Comment by CloudeLecaw** [comment]
This is not possible for the moment for cmi. You need one plugin that ceates one database for one currency. And its Important that they dont sync with vault. Some plugins that usses extra storage for. Money like atms do this. Or plugins where you simply put money on a seperate bank Account away from the normal moneyscore. If you hab sperate databases you can sptead them over your worlds.

What you want i think this is a bit to much at the moment for cmi to be done. Your description sounded like one if these asia Grinding games with extra real money option for shops then Minecraft?.. By the way... This would be forbitten by the eula in Minecraft. Report every mc Server where you see a currency that can be purchased with realmoney - the server will be blocked soon, no matter how big and famous he is ;-)

**Comment by SinsOfSolitude** [comment]
Actually this wouldn't be that hard to make, it would take a bit of time, but it wouldnt be impossible, and yes I know what I'm talking about, used to make my own plugins around 6 years ago.
 I didnt mean this to be used for purchasing any of the special currency with real money, that would be used exactly as I described, you earn low-level currencies (Silver and Gold Coins) in the Bungeecord sub-servers, convert them over to the highest level of currency (Gems) in the subserver shops (which have their DB linked with the Hub servers shop world, as per world syncing would be based off of the world name "Shop" if mysql is also integrated), which is the only currency you can use in the Hub. And that currency would then be used by the players who grind the most, to purchase cosmetic items WITH that currency they earned ingame. It would not be available to purchase for real money, atleast not in my case.
 Breaking eula is kinda not even a factor to bring up on this thread as that can be done in hundreds of other ways, including just vault with the current cmi economy or anything else that manages economy.
 The idea itself came from the real world, how we are able to exchange currencies with different exchange rates. And that points to another thing which I planned for an RPG server, nations with different currencies, each having their own tiers of currency based off their nation.
 Don't really know what you even mean with asia grinding games, unless you're confusing that phrasing with basically every mobile, browser and majority of pc/console games where you're able to purchase special currencies for real money.
 Lastly, not going to sugarcoat this, the way you wrote the second part of your reply, came off insanely assumptive, accusational and just generally made you look like a c*nt, pardon my language. And again about bringing up the Eula, show me exactly where I wrote that any of these currencies could or would be bought for real money or anything else that'd be against the Eula, might make my previous sentence carry slightly less weight then.

**Comment by 456dev** [comment]
firstly, i think its not more if its possible, but if its practical.
to implement this is to basic rewrite the entire economy system and everything it works with (costs of commands (what currency to take from, exchange rates, bla bla bla),
lots of unnecessary complexity for most servers (making it more fragile, and easier to misconfigure)
there is also how many server would even use the feature, as all the effort to fix everything that breaks wouldn't be worth it for the countless other servers its perfectly fine with it currently

TL:DR: slightly flawed i guess, and how many servers would even use it?

p.s. the point about the eula is partially correct, you dont say about it being chared for, but if you can connect to different servers with the same currency then that is against the eula:

> use and sell in-game virtual currencies IF (i) they have no real-world value and that cannot be cashed out, used, **or transferred across free or paid servers**, or into any other currency with real-world value; (ii) you don't give the impression that it comes from or is associated with Mojang, (iii) they don't look like or have similar names to Minecoins or any other official virtual currency, and (iv) anything you sell with the currency follows our monetization guidelines.

(a server being defined as _a single connecting address or IP number_)
from [The Mojang Commercial guidelines](https://account.mojang.com/documents/commercial_guidelines)

**Comment by SinsOfSolitude** [comment]
About implementing this, there are multiple ways of doing this, which can also be a separate add-on to CMI which would replace vault and disable cmi built in economy, or even a secondary plugin which does both of the forementioned things.

To implement it with costs of commands and everything, its easy to do with currency placeholders, as right now the single currency is declared with $. Adding support for it to accept for example [CustomCurrencyNameOrNumber] instead of $. Since exchange rates would be set by the person using the plugin, that wouldn't be any issue, you want a command to be purchasable with X currency, then you just use X where $ should be, and the amount youll figure out based off of your own custom exchange rate.

And I dont really see it being too unnecessarily complex. If the user can set their own currencies and exchange rates, which would require the custom currency placeholder, what other currency is its parent and what the exchange rate to that currency is. Plus a few minor things like naming the currency so if the placeholder is used, it'll show up as what you've set its name as. That's like half a dozen settings per currency, if not less, which could already (as disabled on installation) have a config with pre-set 3 currencies to show the user exactly what to do.

As for the servers that would use it. Potentially a lot of servers that are on bungeecord, prison servers, rpg servers, minigame servers, the list is long. I'm giving the idea as an idea, if it's accepted and done well, the word about it will spread, and people will find interest in it. Also since the minecraft community came alive again like half a year ago or so, a lot of people are trying to get into making servers again, and to make a good server you need a unique idea. With this, you have flexibility on basically every idea that plays part with advanced economy.

And about the Eula, again. If the servers are on bungeecord, you are allowed to use them cross server if all of those servers are on the same IP address. So if i set up a bungeecord network like I've described a few times now, that will all be under a single IP that you can use, all game servers on a single root server. That part of the Eula is a gray area as they haven't defined it well enough, if it's a single IP address the currencies have to be on, then with bungeecord, that is the case, atleast for me. Every server on a single IP yet different port. They have not mentioned anything regarding ports, bungeecord nor syncronizing virtual currency across servers, just that you cant transfer currencies from server A to B, A being for example TheHiveMC and B being for example Mineplex, which have a different IP. I also did not say anything about "transfering" currency to another server, but syncronizing select virtual currencies across select servers on a bungeecord network with mysql, which is completely fine according to the eula.

**Lastly, im writing a "Feature request", which has the desription "Suggest an idea/improvement". >This was an idea<, people should comment on it to help further improve the idea, mention if they would find this useful as well, etc. Thats the whole point of this damn "Feature request" option. Why bash an idea because >you< personally dont like it? Let people who actually do approve of this idea, write their feedback on it, so if Zrips eventually gets to this, then instead of criticism and pointless complaining, he sees how and if people need this. It's also up to him to decide if this is or isn't too complicated to implement either into CMI or as a standalone pay-to-use plugin. Just quit bashing someone elses damn idea because you personally dont see it useful, can't even consider that criticism, just utterly disgusting behaviour.**

**Comment by CloudeLecaw** [comment]
yep your right about this. The Idea of multipble diffent currencies in difffenrt worlds is a nice idea also the exchange but at the moment there is no place for a thing like this in CMI.
but for an addon or extra plugin what would be made to mange this and also to handle all the placeholders
may it would be a nice thing..

What would this mean..
- manage differnt currencies
- manage placeholders to the currencies can be shown in other worlds (hooks)
- manage the other curriens to sync with all other plugin (best way would be if vault can do this because the current most used system to handle currency to commands is vault and not the economiesystem itself)

So CMI would not be the first step where multiple currenies needed to be accepted^^
or like you wrote the plugin can replace also vault + the economie

hmm aboutthe last thing... look arround github you will find enough suggestion where people are telling why it not good or anything no matter if they like it or not.
i never told someone my personal thought about this,,u want? ok...
i diont care about because i wont use this, for me with a signle server without bungeecord a feature like this is worthless ;)
but since this is a feature request i just hope if it ever will be added that it also can be disabled or its simple madew ith a extra plugin like an addon like you descripted it before ;-)

**Comment by SinsOfSolitude** [comment]
Yeah i guess it could be better as an addon plugin. Thanks for giving actual useful feedback to this <3
